packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.tsx

```diff
@@ -0,0 +1,581 @@
+    let isCell = item.type === SwatchColorPickerItemType.Cell;
+    return (
+      <CommandButton
+        { ...item.menuItemButtonProps }
+        id={ id + '-list' + item.index }
+        ref={ 'option' + item.index }
+        key={ id + item.id }
+        data-index={ item.index }
+        data-is-focusable={ true }
+        aria-posinset={ !isCell ? (posInSet && posInSet) : null }
+        aria-setsize={ !isCell ? (setSize && setSize) : null }
+        disabled={ this.props.disabled || item.disabled }
+        className={ css(
+          (isCell ? styles.cell : styles.item),
+          {
+            ['is-selected ' + styles.cellIsSelected]: (isCell && this.state.selectedIndex === item.index),
+            ['is-disabled ' + styles.disabled]: item.disabled && item.disabled
+          },
+        ) }
+        onClick={ () => isCell ? this._onCellClick(item) : this._onMenuItemClick(item) }
+        onMouseEnter={ isCell ? () => this._onItemHoverOrFocused(item, this.props.onCellHovered) : null }
+        onMouseLeave={ isCell ? () => this._clearColors([this.props.onCellHovered]) : null }
+        onFocus={ isCell ? () => this._onItemHoverOrFocused(item, this.props.onCellFocused) : null }
+        role={ isCell ? 'gridcell' : this.props.menuButtonProps ? 'menuitem' : 'button' }
+        aria-selected={ isCell ? (this.state.selectedIndex === item.index ? 'true' : 'false') : null }
+        ariaLabel={ item.label && item.label }
+        title={ item.label && item.label }
+      >
+        { this._onRenderOption(item) }
+      </CommandButton>
+    );
+  }
+
+  /**
+   * Handle hover/focus events for the given item with the given handler
+   * @param item - The item the the event fired against
+   * @param handler - The handler to for the event
+   */
+  private _onItemHoverOrFocused(item: ISwatchColorPickerItemProps, handler?: (id?: string, color?: string) => void) {
+    if (this.props.disabled || (item && item.disabled)) {
+      return;
+    }
+
+    if (handler) {
+      item ? handler(item.id, item.color) : handler();
+    }
+  }
+
+  /**
+   * Handle the click on a cell
+   * @param item - The cell that the click was fired against
+   */
+  private _onCellClick(item: ISwatchColorPickerItemProps) {
+    if (this.props.disabled || item.disabled) {
+      return;
+    }
+
+    let index = item.index;
+
+    // If we have a valid index and it is not already
+    // selected, select it
+    if (index >= 0 && index !== this.state.selectedIndex) {
+      if (this.props.onColorChanged) {
+        this.props.onColorChanged(item.id, item.color);
+      }
+
+      this.setState({
+        selectedIndex: index,
+        expanded: false
+      });
+    } else if (index === this.state.selectedIndex) {
+
+      // The index that got the click was already selected,
+      // clear the selection
+      this._clearColors([this.props.onColorChanged, this.props.onCellHovered, this.props.onCellFocused]);
+
+      this.setState({
+        selectedIndex: -1,
```

Comment:
It's a bit confusing that there are 3 bottom values being used to signal that there isn't a selected index: -1, null, and undefined